### PR TITLE
fabric-ca supports iam

### PR DESCRIPTION
**Pull request by 0xff-dev**
#### Type of change

- New feature

#### Description
fabric-ca support iam.

### CODE REVIEW

**Comment by 0xff-dev** [comment]
@nkwangleiGIT @bjwswang @Abirdcfly 
fabric-ca supports iam service, please review. The following is only the main content to know:
1. **fabric-ca-server**
- Add the iam.enabled parameter
- IAM implements the `Registery` and `User` interfaces
- Add iam authentication function `IAMAuth`

2. **fabric-ca-client**
- Add two parameters `iam.enroll` and `iam.headers=a=a,b=b`
- The client will not use BasicAuth authentication when opening iam

File: cmd/fabric-ca-client/command/config.go [hunk 1]
```diff
@@ -361,6 +368,29 @@ func (c *ClientCmd) createDefaultConfigFile() error {
 	return ioutil.WriteFile(c.cfgFileName, []byte(cfg), 0o755)
 }
 
+func processIAMEnable(cfgHeaders []string, cfg *lib.ClientConfig) error {
```
Comment (bjwswang) [inline cmd/fabric-ca-client/command/config.go:374]: How about use base64 encoding on "username:token" which is same as `basic auth`
```
	header := base64.StdEncoding.EncodeToString([]byte(c.Config.IAM.UserName + ":" + c.Config.IAM.Token))
	request.Header.Set("Authorization", "IAM "+header)
```

File: cmd/fabric-ca-client/command/config.go [hunk 2]
```diff
@@ -343,7 +350,7 @@ func (c *ClientCmd) createDefaultConfigFile() error {
 
 	user := ""
 	var err error
-	if c.requiresUser() {
+	if c.requiresUser() && !c.iamEnroll {
```
Comment (nkwangleiGIT) [inline cmd/fabric-ca-client/command/config.go:356]: should add some comment as tag for changes in old files, for easy merge with upstream later

Comment (0xff-dev) [reply cmd/fabric-ca-client/command/config.go:356]: ok.

File: lib/serverenroll.go [hunk 3]
```diff
@@ -68,7 +68,18 @@ func newReenrollEndpoint(s *Server) *serverEndpoint {
 
 // Handle an enroll request, guarded by basic authentication
 func enrollHandler(ctx *serverRequestContextImpl) (interface{}, error) {
-	id, err := ctx.BasicAuthentication()
+	var (
+		id  string
+		err error
+	)
+	// Goes to different logic depending on whether IAM is on or not.
+	// Goes to different logic depending on whether IAM is on or not.
```
Comment (Abirdcfly) [inline lib/serverenroll.go]: duplicate with line 75

File: lib/server/iam/iam.go [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+func (c *iamClient) GetFilteredUsers(affiliation, types string) (*sqlx.Rows, error) {
+	return nil, errNotSupported
+}
+
+func (c *iamClient) GetAffiliationTree(name string) (*causer.DbTxResult, error) {
+	return nil, errNotSupported
+}
+
+// iamgUser impl User interface
+// func (u *iamUser)
```
Comment (Abirdcfly) [inline lib/server/iam/iam.go]: may delete this line

File: lib/server/iam/iam.go [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+	req.Header.Add("Content-Type", "application/json")
+
+	u := &iamUser{url: c.iamCfg.URL}
+	client := &http.Client{}
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	// if strings.HasPrefix(c.iamCfg.URL, "https") {
```
Comment (bjwswang) [inline lib/server/iam/iam.go:139]: tls handshake with iam still failes?

File: lib/server/iam/iam.go [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+			return nil, err
+		}
+		orgInfo, ok := ua.List[c.organizationID]
+		if !ok {
+			return nil, fmt.Errorf("don't have any organization infomration")
+		}
+		if len(orgInfo.IDs) == 0 {
+			return nil, fmt.Errorf("can't verify user %s", attrNames[0])
+		}
+		if len(attrNames) > 0 {
```
Comment (bjwswang) [inline lib/server/iam/iam.go:220]: why array has been used

**Review by bjwswang — APPROVED** [review]

File: lib/server/iam/iam.go (hunk 6 above)
Comment (0xff-dev) [reply lib/server/iam/iam.go:220]: `attrNames` is mainly for you ldap use, currently iam's enroll needs to deal with peer and order, peer and order names are not too good way to pass into the GetUser call, so the use of `attrName`.

File: lib/server/iam/iam.go (hunk 5 above)
Comment (0xff-dev) [reply lib/server/iam/iam.go:139]: This should require changes to the CA's CRD to mount the IAM-related certificate information when creating the CA.

File: cmd/fabric-ca-client/command/config.go (hunk 1 above)
Comment (0xff-dev) [reply cmd/fabric-ca-client/command/config.go:374]: Good idea. I'll try to change it.

Comment (0xff-dev) [reply cmd/fabric-ca-client/command/config.go:374]: currently revising reenroll related operations and the changes will be submitted together.

**Pull request merged**